libwhisper/WhisperDB.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+	This file is part of cpp-ethereum.
+
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file WhisperDB.cpp
+* @author Vladislav Gluhovsky <vlad@ethdev.com>
+* @date July 2015
+ */
+
+#include <string>
+#include <boost/filesystem.hpp>
+
+#include <libdevcore/Common.h>
+#include <libdevcore/CommonData.h>
+#include <libdevcore/Exceptions.h>
+#include <libdevcore/Log.h>
+#include <libdevcore/SHA3.h>
+#include <libdevcore/FileSystem.h>
+#include "WhisperDB.h"
+
+using namespace std;
+using namespace dev;
+using namespace dev::shh;
+
+WhisperDB::WhisperDB()
+{
+	string path = dev::getDataDir();
+	boost::filesystem::create_directories(path);
+	ldb::Options op;
+	op.create_if_missing = true;
+	op.max_open_files = 256;
+	ldb::DB::Open(op, path + "/whisper", &m_db);
+}
+
+WhisperDB::~WhisperDB()
+{
+	delete m_db;
+}
+
+bool WhisperDB::put(dev::h256 const& _key, string const& _value)
+{
+	string s = _key.hex();
+	string cropped = s.substr(s.size() - 8);
+	leveldb::Status status = m_db->Put(m_writeOptions, s, _value);
```

Comment:
Some bad things are happening here:
- converting the hash to hex format makes it twice as long. LevelDB can store any bytes so put the hash directly.
- You don't have to use std::string as an intermediate representation of the key. LevelDB does not use std::string directly, it uses Slice type similar to our vector_ref. Try to construct the Slice directly from h256.